src/v/datalake/tests/schema_registry_test.cc

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "bytes/iobuf.h"
+#include "datalake/schema_registry.h"
+#include "model/record.h"
+#include "utils/vint.h"
+
+#include <gtest/gtest.h>
+
+#include <variant>
+
+TEST(DatalakeSchemaRegistry, SchemaIdForValidRecord) {
+    iobuf value;
+
+    uint8_t magic = 0;
+    int32_t schema_id = 12;
+    int32_t schema_id_encoded = htobe32(schema_id);
+    std::string payload = "Hello world";
+    value.append(&magic, 1);
+    value.append(reinterpret_cast<uint8_t*>(&schema_id_encoded), 4);
+    value.append(payload.data(), payload.size());
+
+    auto res = datalake::get_value_schema_id(value);
+    ASSERT_TRUE(res.has_value());
+    EXPECT_EQ(res.value().schema_id(), schema_id);
+    EXPECT_EQ(res.value().shared_message_data.size_bytes(), payload.size());
+}
+
+TEST(DatalakeSchemaRegistry, SchemaIdForShortRecord) {
+    iobuf value;
+
+    uint8_t magic = 0;
+    int32_t schema_id = 12;
+    int32_t schema_id_encoded = htobe32(schema_id);
+    value.append(&magic, 1);
+    // Only adding 3 bytes here instead of 4 to generate an invalid record.
+    value.append(reinterpret_cast<uint8_t*>(&schema_id_encoded), 3);
+
+    auto res = datalake::get_value_schema_id(value);
+    ASSERT_TRUE(res.has_error());
+    EXPECT_EQ(res.error(), datalake::get_schema_error::not_enough_bytes);
+}
+
+TEST(DatalakeSchemaRegistry, SchemaIdForBadMagic) {
+    iobuf value;
+
+    uint8_t magic = 5; // Invalid magic
+    int32_t schema_id = 12;
+    int32_t schema_id_encoded = htobe32(schema_id);
+    value.append(&magic, 1);
+    value.append(reinterpret_cast<uint8_t*>(&schema_id_encoded), 4);
+    auto res = datalake::get_value_schema_id(value);
+    ASSERT_TRUE(res.has_error());
+    EXPECT_EQ(res.error(), datalake::get_schema_error::no_schema_id);
+}
+
+TEST(DatalakeSchemaRegistry, GetProtoOffsetsOk) {
+    iobuf value;
+
+    uint8_t magic = 0; // Invalid magic
+    uint32_t schema_id = 12;
+    int32_t schema_id_encoded = htobe32(schema_id);
+    std::string payload = "Hello world";
+    value.append(&magic, 1);
+    value.append(reinterpret_cast<uint8_t*>(&schema_id_encoded), 4);
+
+    uint8_t proto_msg_count = 9;
+    std::array<uint8_t, 16> encoded;
+    size_t encoded_size = vint::serialize(proto_msg_count, encoded.data());
```

Comment:
nit: `vint::to_bytes` is probably simpler and will make this a bit more readable.